### ISSUE TITLE
How to use ConvLSTM after Conv layers?

### ISSUE THREAD

**Issue by tyshiwo** [agethen/ConvLSTM-for-Caffe#6]
Hi,

I wonder can I use the outputs of several Conv layers (e.g., T Conv layers) as the input of this ConvLSTM layer (here, T can be regarded as the time-steps)? In such case, do I need to prepare other inputs for the ConvLSTM layer? In your example, there are lots of inputs, which seems very complex. 

Any suggestions would be appreciated. Thanks!

**Comment by agethen** [comment]
If you have multiple inputs, you could just concatenate them. Caffe has a "Concatenate" layer:
```
layer {
  name: "concatenate"
  type: "Concatenate"
  bottom: "in1"
  bottom: "in2"
  bottom: "in3"

  top: "out"

  concatenate_param {
    axis: <xxxx>
  }
}
```
All inputs must have the same dimensions of course, except on the concatenation axis.
Is that what you want?

**Comment by tyshiwo** [comment]
Thanks for your response.

Actually, my question is what should I do to format those feature maps from the previous Conv layers to be the input of the ConvLSTM layer. 

Suppose each of the conv layer generates feature maps with size to be N*C*H*W. Using Concat layer in caffe, we may have (TN)*C*H*W with axis=0, or N*(TC)*H*W with axis=1. Is the size (TN)*C*H*W or N*(TC)*H*W good for ConvLSTM? Since in your case, you have 4 input blobs, e.g., the following case:

layer {
  name: "encode1"
  type: "ConvLSTM"

  bottom: "input"     # Input features x
  bottom: "seq_enc"   # Sequence markers

  bottom: "dummy"   # Dummy input h (required by `expose_hidden: true`)
  bottom: "dummy"   # Dummy input c

  top: "encode1"
  top: "encode1_h"    # Final hidden state (Shape: 1 x N x C x H x W)
  top: "encode1_c"    # Final cell state

  # This is a parameter from Jeff Donahues implementation.
  # It allows us to expose `encode1_h` and `encode1_c`.
  recurrent_param {
    expose_hidden: true
  }

  lstm_debug_param{
    # These two are actually default values and do not need to be set.
    # num_axes_hadamard: 2 implies that parameters to be learnt are a 2D tensor
    # axis_hadamard: 3 implies where to apply the product (
    # Therefore, given TxNxCxHxW, the parameters have shape H x W.
    axis_hadamard: 3      
    num_axes_hadamard: 2  
  }

  # Conv Layer specification
  lstm_conv_param {
    num_output: 256
    kernel_size: 5
    pad: 2                  # Padding is required! All conv. kernels need to result in the same H x W shape.

    weight_filler {
      # Do not use xavier or other automatic methods on 5D tensors, this may cause bugs.
      # Rather use uniform or gaussian!
      # Xavier: sqrt( 3 / fan_in), fan_in := input_count / N, input_count := 10*16*16*16*16, N := 16
      # Technically, we would need to specify different for type: "input" and type: "hidden"...
      type: "gaussian"
      std: 0.01
    }
    bias_filler {
      type: "constant"
      value: 0
    }
  }
}

**Comment by tyshiwo** [comment]
Thanks for your response.

Actually, my question is what should I do to format those feature maps from the previous Conv layers to be the input of the ConvLSTM layer.

Suppose each of the conv layer generates feature maps with size to be NCHW. Using Concat layer in caffe, we may have (TN)xCxHxW with axis=0, or 

Nx(TC)xHxW with axis=1. Is the size (TN)xCxHxW or Nx(TC)xHxW good for ConvLSTM? Since in your case, you have 4 input blobs, e.g., the following 

case:

layer {
  name: "encode1"
  type: "ConvLSTM"

  bottom: "input"     # Input features x
  bottom: "seq_enc"   # Sequence markers

  bottom: "dummy"   # Dummy input h (required by `expose_hidden: true`)
  bottom: "dummy"   # Dummy input c

  top: "encode1"
  top: "encode1_h"    # Final hidden state (Shape: 1 x N x C x H x W)
  top: "encode1_c"    # Final cell state

  recurrent_param {
    expose_hidden: true
  }

  lstm_debug_param{
    axis_hadamard: 3      
    num_axes_hadamard: 2  
  }

  lstm_conv_param {
    num_output: 256
    kernel_size: 5
    pad: 2               

    weight_filler {
      type: "gaussian"
      std: 0.01
    }
    bias_filler {
      type: "constant"
      value: 0
    }
  }
}

**Comment by tyshiwo** [comment]
Thanks for your response.

Actually, my question is what should I do to format those feature maps from the previous Conv layers to be the input of the ConvLSTM layer.

Suppose each of the conv layer generates feature maps with size to be NCHW. Using Concat layer in caffe, we may have (TN)xCxHxW with axis=0, or Nx(TC)xHxW with axis=1. Is the size (TN)xCxHxW or Nx(TC)xHxW good for ConvLSTM? Since in your case, you have 4 input blobs, e.g., the following case:

layer {
  name: "encode1"
  type: "ConvLSTM"

  bottom: "input"     # Input features x
  bottom: "seq_enc"   # Sequence markers

  bottom: "dummy"   # Dummy input h (required by `expose_hidden: true`)
  bottom: "dummy"   # Dummy input c

  top: "encode1"
  top: "encode1_h"    # Final hidden state (Shape: 1 x N x C x H x W)
  top: "encode1_c"    # Final cell state

  recurrent_param {
    expose_hidden: true
  }

  lstm_debug_param{
    axis_hadamard: 3      
    num_axes_hadamard: 2  
  }

  lstm_conv_param {
    num_output: 256
    kernel_size: 5
    pad: 2               

    weight_filler {
      type: "gaussian"
      std: 0.01
    }
    bias_filler {
      type: "constant"
      value: 0
    }
  }
}

**Comment by agethen** [comment]
Ah, I see!
The correct concatenation is along axis = 0, so T*N x C x H x W. After concatenation, you then reshape to T x N x C x H x W.

Btw, LSTM by Jeff Donahue and my ConvLSTM only require two input blobs (the first two). In the example that you quote, I showed how to expose the cell state C and hidden state H (for example with the purpose to input it into another layer). If you do not need C or H, feel free to set ``expose_hidden: false`` and remove the corresponding two input and output blobs (in this example: "dummy", "encode1_h", "encode1_c").

**Comment by tyshiwo** [comment]
I see. Thanks!

Another simple question is what is the effect of the second input "seq" containing 0 or 1 in this example?

**Comment by agethen** [comment]
The sequence indicators are just the same as for the default LSTM. 
At time t, the previous hidden- and cell states are multiplied by that value, allowing you to "reset" the LSTM at a certain time (by specifying a 0).

**Comment by tyshiwo** [comment]
I see. I have successfully trained your convLSTM version for a regression task. The training is good and the testing using command (e.g., $CAFFE test --model=XXX --weights=XXX) is also ok.  My task is to do regression but using command I can just see the loss.

To achieve the regression outputs, I also use matlab to test the model. However, I met the following error:
![image](https://cloud.githubusercontent.com/assets/19333875/21761316/1e88777c-d68d-11e6-94ad-d97c7fa1c400.png)
In this case, the timestep is set to be 10. It seems the size of a blob is larger than the maximum size. I can't figure out why it is ok when using command but fails when using matlab interface? I also tried a smaller timestep (i.e., 5), the error still exists:
![image](https://cloud.githubusercontent.com/assets/19333875/21761372/6f1eec3e-d68d-11e6-9a99-b64e1c1a635b.png)
Have you ever met this problem before? Any suggestion will be appreciated! Thanks!

**Comment by agethen** [comment]
Hey!

Well, that is really difficult to say without having any details. So, you just create a network with the same prototxt as for the C++ standalone? Can you show some relevant part? What are the dimensions (kernel,input) for the ConvLSTM?
To be honest, I am not particularly familiar with the Matlab wrapper though :/

**Comment by tyshiwo** [comment]
Hi,

Here is some part of my prototxt, which concatenates 5 previous convs' outputs. The dimensions of kernel and input are 32x3x3 and 5x32x32x31x31, respectively. The batchsize is set to 32 and the input patch is with the size of 31x31.

layer {
    name: "concat_conv"
    type: "Concat"
    bottom: "conv01"
    bottom: "conv02"
    bottom: "conv03"
    bottom: "conv04"
    bottom: "conv05"
    top: "concat_conv"
    concat_param {
        axis: 0
    }
}
layer {
    name: "reshape_conv"
    type: "Reshape"
    bottom: "concat_conv"
    top: "reshape_conv"
    reshape_param {
      shape {
        dim: 5
        dim: 32
        dim: 32
        dim: 31
        dim: 31
      }
    }
}
layer {
    name: "bn_convlstm01"
    type: "BatchNorm"
    bottom: "reshape_conv"
    top: "bn_convlstm01"
    param {
        lr_mult: 0
    }
    param {
        lr_mult: 0
    }
    param {
        lr_mult: 0
    }
}
layer {
    name: "scale_convlstm01"
    type: "Scale"
    bottom: "bn_convlstm01"
    top: "bn_convlstm01"
    scale_param {
        bias_term: true
    }
}
layer {
    name: "relulstm01"
    type: "ReLU"
    bottom: "bn_convlstm01"
    top: "bn_convlstm01"
}
layer {
    name: "convlstm01"
    type: "ConvLSTM"
    bottom: "bn_convlstm01"
    bottom: "seq"
    top: "convlstm01"
    recurrent_param {
      expose_hidden: false
    }
    lstm_debug_param {
      axis_hadamard: 3
      num_axes_hadamard: 2
    }
    lstm_conv_param {
        num_output: 32
        kernel_size: 3
        stride: 1
        pad: 1
        weight_filler {
            type: "gaussian"
            std: 1.000000e-02
        }
        bias_filler {
            type: "constant"
            value: 0
        }
    }
}
layer {
    name: "slice_convlstm01"
    type: "Slice"
    bottom: "convlstm01"
    top: "convlstm01_discard"
    top: "convlstm01_slice"
    slice_param {
        axis: 0
        slice_point: 9
    }
}
layer {
    name: "silence_convlstm01"
    type: "Silence"
    bottom: "convlstm01_discard"
}
layer {
    name: "reshape_convlstm"
    type: "Reshape"
    bottom: "convlstm01_slice"
    top: "reshape_convlstm"
    reshape_param {
      shape {
        dim: 32
        dim: 32
        dim: 31
        dim: 31
      }
    }
}

I think the problem may be the format of my deploy file. Usually, I used matlab wrapper to do testing. Since in the previous models, I only have one input ("data", no "seq"), the input in the deploy file would be:
input: "data"
input_dim: 1
input_dim: 1
input_dim: 31
input_dim: 31

And the testing is okay. However, in the convLSTM version, since there's another input "seq", I wrote:
input: "data"
input_dim: 1
input_dim: 1
input_dim: 31
input_dim: 31
input: "seq"
input_dim: 5
input_dim: 1

This gives an error. I'm not sure whether such format is valid or not.

**Comment by tyshiwo** [comment]
I modified caffe to get regression outputs through command. But I met another problem: I used a fixed input size during training (e.g., 96x96), my testing images are with arbitrary sizes (e.g., 128x128, 256x288). I thought convLSTM could handle this problem, but when I used a different input size for testing, it told me:
![image](https://cloud.githubusercontent.com/assets/19333875/21792166/c47cdd2c-d722-11e6-9485-823827e909be.png)
Does it mean the convLSTM is not like the fully convolutional network that are insensitive to the input size?

**Comment by agethen** [comment]
Concerning the latter problem, this is due to the Hadamard term. It takes on the size of a single channel of the hidden state. 
You can disable that by specifying `disable_hadamard: true` inside the lstm_debug_param.
(However, you might not be able to copy weights between networks that have different settings...)

Concerning the former problem, I will look into your prototxt later today, maybe I find something.

**Comment by agethen** [comment]
So, concering the protofile:
First, layer "slice_convlstm01" is trying to slice at T=9, but from what I understand you only have T=5. So you might want to remove or modify that layer.

Then, you do not really need to write a deploy.prototxt file, or? You can just reuse the train.prototxt, and set the phase (caffe.TRAIN/caffe.TEST) correctly. At least that works in the python wrapper :)
Though, looks like the data blobs of the deploy file you showed me only have batch size=1.

I am still not sure of the reason. In my experience, this usually happened when the input to ConvLSTM had a wrong shape. So I would suspect either your deploy file or that matlab is treating shapes differently...

**Comment by tyshiwo** [comment]
Yup, "T=9" actually is a typo, I have already modified it.

The reason why I wrote a deploy file is in my task (e.g., image restoration), the test images can be with different sizes, directly using train.prototxt only deals with one input size. Therefore, usually I use matlab wrapper to send one image into the deploy file, which can be modified dynamically.

Anyway, thanks!

**Comment by tyshiwo** [comment]
Hi,

Sorry to disturb you again. I tried disable_hadamard: true or false, and found both cases fail when the batch sizes are different between training and testing using matlab wrapper (it is good when the batch sizes are kept the same between training and testing). Is that normal? 

Thanks!

**Comment by agethen** [comment]
Hey!
It definitely should not cause any problems. Did you remember to adjust all reshape layers accordingly?
